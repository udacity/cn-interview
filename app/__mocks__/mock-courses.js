export const courses = [
  {
    key: 1,
    img: 'mock-url-1',
    title: 'mock-title-1',
    level: '初级',
    tracks: ['数据科学'],
    subtitle: 'mock-subtitle-1',
    short_summary: 'mock-short-summary-1',
    projects: ['project-1', 'project-2', 'project-3'],
  },
  {
    key: 2,
    img: 'mock-url-2',
    title: 'mock-title-2',
    level: '中级',
    tracks: ['iOS'],
    subtitle: 'mock-subtitle-2',
    short_summary: 'mock-short-summary-2',
    projects: ['project-1'],
  },
  {
    key: 3,
    img: 'mock-url-3',
    title: 'mock-title-3',
    level: '高级',
    tracks: ['网站开发'],
    subtitle: 'mock-subtitle-3',
    short_summary: 'mock-short-summary-3',
    projects: [],
  },
];
