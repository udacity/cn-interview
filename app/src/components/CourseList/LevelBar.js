import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Square from 'react-icons/lib/fa/stop';
import { getIconArr } from '../../utils/helper';

export default class LevelBar extends Component {
  getLevelIcons = iconArr => {
    const levelIcons = iconArr.map((each, index) => {
      if (each) {
        return <Square key={index} size={16} color="#666" />;
      }
      return <Square key={index} size={16} color="#dcdcdc" />;
    });
    return levelIcons;
  };

  static propTypes = {
    level: PropTypes.string,
  };

  render() {
    const { level } = this.props;
    const iconArr = getIconArr(level);
    const levelIcons = this.getLevelIcons(iconArr);

    return (
      <div className="level-bar vertical-center">
        <div className="level-icons vertical-center">{levelIcons}</div>
        <span>{level}</span>
      </div>
    );
  }
}
