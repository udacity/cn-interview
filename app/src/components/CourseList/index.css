.course-list-item {
  padding-top: 20px;
}

.course-list-projects {
  margin-bottom: 20px;
}

.level-icons {
  margin-right: 20px;
}

.level-bar {
  font-size: 16px;
}