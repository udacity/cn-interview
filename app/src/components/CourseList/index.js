import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Image } from 'react-bootstrap';
import LevelBar from './LevelBar';
import './index.css';

export class Item extends Component {
  static propTypes = {
    image: PropTypes.string,
    title: PropTypes.string,
    level: PropTypes.string,
    subtitle: PropTypes.string,
    short_summary: PropTypes.string,
    projects: PropTypes.array,
  };

  render() {
    const {
      image,
      title,
      level,
      subtitle,
      short_summary: shortSummary,
      projects,
    } = this.props;
    const projectNum = projects.length;
    return (
      <Row className="course-list-item">
        <Col md={3}>
          <Image className="img-fluid" src={image} />
        </Col>
        <Col md={9}>
          <Row>
            <Col md={8}>
              <h3 className="h-title">{title}</h3>
            </Col>
            <Col md={4}>
              <LevelBar level={level} />
            </Col>
          </Row>
          <h4 className="h-light course-list-projects">
            {projectNum ? `${projectNum} 个` : '没有'}项目
          </h4>
          <h3 className="h-bold">{subtitle}</h3>
          <p className="p-normal">{shortSummary}</p>
        </Col>
      </Row>
    );
  }
}

export default class CourseList extends Component {
  static propTypes = {
    courses: PropTypes.array,
  };

  render() {
    const { courses = [] } = this.props;
    const listItems = courses.map(course => (
      <Item {...course} key={course.key} />
    ));

    return <div className="course-list">{listItems}</div>;
  }
}
