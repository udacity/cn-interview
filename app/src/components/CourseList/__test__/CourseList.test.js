import React from 'react';
import { mount } from 'enzyme';
import CourseList from '..';
import { courses } from 'mock-courses';

it('renders correctly', () => {
  const wrapper = mount(<CourseList courses={courses} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders without crashing when courses is empty', () => {
  mount(<CourseList courses={[]} />);
});

it('should have correct project message', () => {
  const wrapper = mount(<CourseList courses={[...courses]} />);
  expect(
    wrapper
      .find('h4')
      .at(1)
      .text()
  ).toEqual('1 个项目');
  expect(
    wrapper
      .find('h4')
      .at(2)
      .text()
  ).toEqual('没有项目');
});
