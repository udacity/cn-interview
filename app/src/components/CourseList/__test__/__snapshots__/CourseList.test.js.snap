// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<CourseList
  courses={
    Array [
      Object {
        "img": "mock-url-1",
        "key": 1,
        "level": "初级",
        "projects": Array [
          "project-1",
          "project-2",
          "project-3",
        ],
        "short_summary": "mock-short-summary-1",
        "subtitle": "mock-subtitle-1",
        "title": "mock-title-1",
        "tracks": Array [
          "数据科学",
        ],
      },
      Object {
        "img": "mock-url-2",
        "key": 2,
        "level": "中级",
        "projects": Array [
          "project-1",
        ],
        "short_summary": "mock-short-summary-2",
        "subtitle": "mock-subtitle-2",
        "title": "mock-title-2",
        "tracks": Array [
          "iOS",
        ],
      },
      Object {
        "img": "mock-url-3",
        "key": 3,
        "level": "高级",
        "projects": Array [],
        "short_summary": "mock-short-summary-3",
        "subtitle": "mock-subtitle-3",
        "title": "mock-title-3",
        "tracks": Array [
          "网站开发",
        ],
      },
    ]
  }
>
  <div
    className="course-list"
  >
    <Item
      img="mock-url-1"
      key="1"
      level="初级"
      projects={
        Array [
          "project-1",
          "project-2",
          "project-3",
        ]
      }
      short_summary="mock-short-summary-1"
      subtitle="mock-subtitle-1"
      title="mock-title-1"
      tracks={
        Array [
          "数据科学",
        ]
      }
    >
      <Row
        bsClass="row"
        className="course-list-item"
        componentClass="div"
      >
        <div
          className="course-list-item row"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={3}
          >
            <div
              className="col-md-3"
            >
              <Image
                bsClass="img"
                circle={false}
                className="img-fluid"
                responsive={false}
                rounded={false}
                thumbnail={false}
              >
                <img
                  className="img-fluid"
                />
              </Image>
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <div
              className="col-md-9"
            >
              <Row
                bsClass="row"
                componentClass="div"
              >
                <div
                  className="row"
                >
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={8}
                  >
                    <div
                      className="col-md-8"
                    >
                      <h3
                        className="h-title"
                      >
                        mock-title-1
                      </h3>
                    </div>
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={4}
                  >
                    <div
                      className="col-md-4"
                    >
                      <LevelBar
                        level="初级"
                      >
                        <div
                          className="level-bar vertical-center"
                        >
                          <div
                            className="level-icons vertical-center"
                          >
                            <FaStop
                              color="#666"
                              key="0"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#dcdcdc"
                              key="1"
                              size={16}
                            >
                              <IconBase
                                color="#dcdcdc"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#dcdcdc",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#dcdcdc"
                              key="2"
                              size={16}
                            >
                              <IconBase
                                color="#dcdcdc"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#dcdcdc",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                          </div>
                          <span>
                            初级
                          </span>
                        </div>
                      </LevelBar>
                    </div>
                  </Col>
                </div>
              </Row>
              <h4
                className="h-light course-list-projects"
              >
                3 个
                项目
              </h4>
              <h3
                className="h-bold"
              >
                mock-subtitle-1
              </h3>
              <p
                className="p-normal"
              >
                mock-short-summary-1
              </p>
            </div>
          </Col>
        </div>
      </Row>
    </Item>
    <Item
      img="mock-url-2"
      key="2"
      level="中级"
      projects={
        Array [
          "project-1",
        ]
      }
      short_summary="mock-short-summary-2"
      subtitle="mock-subtitle-2"
      title="mock-title-2"
      tracks={
        Array [
          "iOS",
        ]
      }
    >
      <Row
        bsClass="row"
        className="course-list-item"
        componentClass="div"
      >
        <div
          className="course-list-item row"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={3}
          >
            <div
              className="col-md-3"
            >
              <Image
                bsClass="img"
                circle={false}
                className="img-fluid"
                responsive={false}
                rounded={false}
                thumbnail={false}
              >
                <img
                  className="img-fluid"
                />
              </Image>
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <div
              className="col-md-9"
            >
              <Row
                bsClass="row"
                componentClass="div"
              >
                <div
                  className="row"
                >
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={8}
                  >
                    <div
                      className="col-md-8"
                    >
                      <h3
                        className="h-title"
                      >
                        mock-title-2
                      </h3>
                    </div>
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={4}
                  >
                    <div
                      className="col-md-4"
                    >
                      <LevelBar
                        level="中级"
                      >
                        <div
                          className="level-bar vertical-center"
                        >
                          <div
                            className="level-icons vertical-center"
                          >
                            <FaStop
                              color="#666"
                              key="0"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#666"
                              key="1"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#dcdcdc"
                              key="2"
                              size={16}
                            >
                              <IconBase
                                color="#dcdcdc"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#dcdcdc",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                          </div>
                          <span>
                            中级
                          </span>
                        </div>
                      </LevelBar>
                    </div>
                  </Col>
                </div>
              </Row>
              <h4
                className="h-light course-list-projects"
              >
                1 个
                项目
              </h4>
              <h3
                className="h-bold"
              >
                mock-subtitle-2
              </h3>
              <p
                className="p-normal"
              >
                mock-short-summary-2
              </p>
            </div>
          </Col>
        </div>
      </Row>
    </Item>
    <Item
      img="mock-url-3"
      key="3"
      level="高级"
      projects={Array []}
      short_summary="mock-short-summary-3"
      subtitle="mock-subtitle-3"
      title="mock-title-3"
      tracks={
        Array [
          "网站开发",
        ]
      }
    >
      <Row
        bsClass="row"
        className="course-list-item"
        componentClass="div"
      >
        <div
          className="course-list-item row"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={3}
          >
            <div
              className="col-md-3"
            >
              <Image
                bsClass="img"
                circle={false}
                className="img-fluid"
                responsive={false}
                rounded={false}
                thumbnail={false}
              >
                <img
                  className="img-fluid"
                />
              </Image>
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <div
              className="col-md-9"
            >
              <Row
                bsClass="row"
                componentClass="div"
              >
                <div
                  className="row"
                >
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={8}
                  >
                    <div
                      className="col-md-8"
                    >
                      <h3
                        className="h-title"
                      >
                        mock-title-3
                      </h3>
                    </div>
                  </Col>
                  <Col
                    bsClass="col"
                    componentClass="div"
                    md={4}
                  >
                    <div
                      className="col-md-4"
                    >
                      <LevelBar
                        level="高级"
                      >
                        <div
                          className="level-bar vertical-center"
                        >
                          <div
                            className="level-icons vertical-center"
                          >
                            <FaStop
                              color="#666"
                              key="0"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#666"
                              key="1"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                            <FaStop
                              color="#666"
                              key="2"
                              size={16}
                            >
                              <IconBase
                                color="#666"
                                size={16}
                                viewBox="0 0 40 40"
                              >
                                <svg
                                  fill="currentColor"
                                  height={16}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#666",
                                      "verticalAlign": "middle",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                  width={16}
                                >
                                  <g>
                                    <path
                                      d="m37.3 4.3v31.4q0 0.6-0.4 1t-1 0.4h-31.5q-0.6 0-1-0.4t-0.4-1v-31.4q0-0.6 0.4-1t1-0.4h31.5q0.5 0 1 0.4t0.4 1z"
                                    />
                                  </g>
                                </svg>
                              </IconBase>
                            </FaStop>
                          </div>
                          <span>
                            高级
                          </span>
                        </div>
                      </LevelBar>
                    </div>
                  </Col>
                </div>
              </Row>
              <h4
                className="h-light course-list-projects"
              >
                没有
                项目
              </h4>
              <h3
                className="h-bold"
              >
                mock-subtitle-3
              </h3>
              <p
                className="p-normal"
              >
                mock-short-summary-3
              </p>
            </div>
          </Col>
        </div>
      </Row>
    </Item>
  </div>
</CourseList>
`;
